common/validator_dir/src/builder.rs

```diff
@@ -215,26 +222,35 @@ impl<'a> Builder<'a> {
                 }
             }
 
-            // Only the withdrawal keystore if explicitly required.
-            if self.store_withdrawal_keystore {
-                // Write the withdrawal password to file.
-                write_password_to_file(
-                    self.password_dir
-                        .join(withdrawal_keypair.pk.to_hex_string()),
-                    withdrawal_password.as_bytes(),
-                )?;
-
-                // Write the withdrawal keystore to file.
-                write_keystore_to_file(dir.join(WITHDRAWAL_KEYSTORE_FILE), &withdrawal_keystore)?;
+            if self.password_dir.is_none() && self.store_withdrawal_keystore {
+                return Err(Error::MissingPasswordDir);
+            }
+
+            if let Some(password_dir) = self.password_dir.as_ref() {
+                // Only the withdrawal keystore if explicitly required.
+                if self.store_withdrawal_keystore {
+                    // Write the withdrawal password to file.
+                    write_password_to_file(
+                        password_dir.join(withdrawal_keypair.pk.to_hex_string()),
+                        withdrawal_password.as_bytes(),
+                    )?;
+
+                    // Write the withdrawal keystore to file.
+                    write_keystore_to_file(
+                        dir.join(WITHDRAWAL_KEYSTORE_FILE),
+                        &withdrawal_keystore,
+                    )?;
+                }
             }
         }
 
-        // Write the voting password to file.
-        write_password_to_file(
-            self.password_dir
-                .join(format!("0x{}", voting_keystore.pubkey())),
-            voting_password.as_bytes(),
-        )?;
+        if let Some(password_dir) = self.password_dir.as_ref() {
+            // Write the voting password to file.
+            write_password_to_file(
+                password_dir.join(format!("0x{}", voting_keystore.pubkey())),
+                voting_password.as_bytes(),
+            )?;
+        }
```

Comment:
Might be simpler with something like this:

```
            if let Some(password_dir) = self.password_dir.as_ref() {
                // Only the withdrawal keystore if explicitly required.
                if self.store_withdrawal_keystore {
                    // Write the withdrawal password to file.
                    write_password_to_file(
                        password_dir.join(withdrawal_keypair.pk.to_hex_string()),
                        withdrawal_password.as_bytes(),
                    )?;

                    // Write the withdrawal keystore to file.
                    write_keystore_to_file(
                        dir.join(WITHDRAWAL_KEYSTORE_FILE),
                        &withdrawal_keystore,
                    )?;
                }
                // Write the voting password to file.
                write_password_to_file(
                    password_dir.join(format!("0x{}", voting_keystore.pubkey())),
                    voting_password.as_bytes(),
                )?;
            } else if self.store_withdrawal_keystore {
                return Err(Error::MissingPasswordDir);
            }
```

Reply:
I'm not sure this will still store the voting keystore if the withdrawal keystore is not provided.